### PR TITLE
Patient-request audit logging.

### PR DESCRIPTION
**Pull request by benwarfield-usds**
Added a request interceptor for the non-graphql `@Controller` beans, and set it up to log the important parts of each request in postHandle, after setting up a log-friendly requestId in preHandle.

## Related Issue or Background Info

- #522 

## Changes Proposed

- add a new field to ApiAuditEvent to hold HTTP response status (defaulting to 200 since graphql is like that)
- add an interceptor that wraps non-graphql requests and logs them if there is a PatientLink associated
- add a controller advice that intercepts exceptional exits from REST controllers and logs the audit event if needed

## TODO

- [x] a better way to determine if a request should be logged than "do we have a user?"
- [x] make sure that we catch expired-link requests
- [x] add assertions to most of the PXP tests to validate that the correct audit events are created
- [x] validate the assertions made in our AU-5 response

## Additional Information

- we do not log any request parameters other than the patient link itself, because we only need to know who was acting and what kind of action they were attempting, not exactly what they did (we can't readily distinguish between filtering parameters and mutating parameters in graphql, so we log more than we strictly need to, there)

## Follow-ups

- a more graceful handling of log failures in graphql
- some test infrastructure rearrangement
- fix double-logging of REST errors
- add RequestId header to graphql responses

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      HttpServletResponse response,
+      Object handler,
+      ModelAndView modelAndView)
+      throws Exception {
+    PatientLink patientLink = _context.getPatientLink();
+    LOG.debug("Closing out request. Patient link is {}", patientLink);
+    if (patientLink != null) { // this is noooot acceptable
+      String requestId = MDC.get(GraphQLLoggingHelpers.GRAPHQL_QUERY_MDC_KEY);
+      _auditService.logRestEvent(requestId, request, _context.getOrganization(), patientLink);
+    }
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:56]: Should this maybe log a stern warning and capture an audit event with a missing patient link if `patientLink` is `null`?

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws Exception {
+    LOG.debug(
+        "Pre-handling a method={} request for uri={} using handler={}",
+        request.getMethod(),
+        request.getRequestURI(),
+        handler);
+    String requestId = UUID.randomUUID().toString();
+    MDC.put(GraphQLLoggingHelpers.GRAPHQL_QUERY_MDC_KEY, requestId);
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:40]: Should this be added a response header here? Not sure if that would make Spring write the headers immediately, which would be not good.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiAuditEvent.java [hunk 3]
```diff
@@ -125,6 +126,21 @@ public ApiAuditEvent(
...
   }
 
+  /** Constructor for REST (patient-experience) requests */
+  public ApiAuditEvent(
+      String requestId,
+      HttpRequestDetails httpRequestDetails,
+      ApiUser user,
+      Organization organization,
+      PatientLink patientLink) {
+    super();
```
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/ApiAuditEvent.java:136]: dammit

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java (hunk 1 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:56]: That seems like an important question. My basic thinking is

1) this is to capture actions by patients
2) absent a PatientLink we cannot identify the patient
3) AU-5 says we bail if we can't record things
4) now therefore, the absence of a validated PatientLink should be a fatal error and we should kick back some 4xx error and call it a day.

This is slightly complicated by 
1. the fact that we are sitting _above_ the authentication layer, so something can come back to us with no patient link but a 403 error already, which is something we neither want to change or log; and
2. the recent introduction of legitimately anonymous endpoints (the registration form).

My solution for the second one is to explicitly identify anonymous-OK endpoints, either by explicit whitelist or by a marker interface or annotation (the last one being my general preference). With that done, anybody who has no PatientLink gets the greater of 400 and whatever they already had, and I think it more or less works out.

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java (hunk 2 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:40]: I'm fairly sure that we can add a response header here—if not, we can certainly add it in postHandle. Do we want to do the same for graphql responses?

Comment (jeskew-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:42]: I thought we were already setting one on graphql responses, but it looks like I was thinking of a request header being set by the Apollo client.

Let's add the header to both response types to aid in debugging.

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java (hunk 1 above)
Comment (jeskew-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java]: I think for anonymous endpoints and failed authentication attempts, we can't identify the requestor. We only need to bail if writing an audit event fails, not if there's some information which can't be known for a the action being recorded.

Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java]: Oh, I suppose we do probably want to record the an audit event for the anonymous user, at that. Too bad I made api_user_id a non-null field! &lt;sight&gt;

Logging authentication failures introduces an asymmetry, though: we don't log 403 or 401 responses on the graphql endpoint, because we never see them. But then, Okta probably does log auth failures on their side. So should we be logging those as well, here?

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java (hunk 2 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:42]: Hmm. This may be a mildly vexing proposition on the graphql side, but I think it's doable. And I probably just spent somewhat longer than I should have finding out.

**Comment by benwarfield-usds** [comment]
This still needs tests for a particular set of edge cases, but the main code flow and the bulk of tests are in place.

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java (hunk 2 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java]: Going to make adding it to graphql a separate PR.

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  @Override
+  public void afterCompletion(
+      HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex)
+      throws Exception {
+    LOG.trace("Final logging cleanup step.");
+    MDC.remove(LoggingConstants.REQUEST_ID_MDC_KEY);
+  }
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:73]: Spring n00b question: if logging were happening here instead of in `postHandle`, would you need a separate `ControllerAdvice` logger to intercept exceptions? Not sure if there are some conditions under which Spring would not invoke `onComplete`.

Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:73]: Not a bad question at all, actually—I spent a chunk of the weekend trying to get a good answer on it. To the best of my belief, you would not need a separate ControllerAdvice _but_ you would also not be able to fully enforce the AU-5 restriction on responses going out without logging. By the time we get here, the response is committed, so we can't change the content if the audit log fails. Since I am operating under the assumption that I can't hand-wave that away, I went with the ControllerAdvice.

I could of course be totally wrong in that belief—I didn't experiment!—but I did research it at least a bit, and this is where I landed.

**Review by timothybest-usds — APPROVED** [review]

File: backend/src/test/java/gov/cdc/usds/simplereport/api/AuditLoggingFailuresTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    assertEquals(List.of("addTestResult"), event.getGraphqlErrorPaths());
+  }
+
+  @Test
+  void graphqlQuery_auditFailure_noDataReturned() {
+    when(_auditRepo.save(_eventCaptor.capture()))
+        .thenThrow(new IllegalArgumentException("naughty naughty"));
+    useOrgUserAllFacilityAccess();
+    ObjectNode args = patientArgs().put("symptoms", "{}").put("noSymptoms", true);
+    assertThrows(NullPointerException.class, () -> runQuery("update-time-of-test", args));
```
Comment (jeskew-gov) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/AuditLoggingFailuresTest.java:89]: Not really following this one. Shouldn't it have thrown an IllegalArgumentException?

File: backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    assertEquals(errorPaths, event.getGraphqlErrorPaths(), "Query paths with errors");
+    return event;
+  }
+
+  /** REST audit event checker: includes "this is not graphql" checks" */
+  protected ApiAuditEvent assertLastAuditEntry(
+      HttpStatus status, String requestUri, String requestId) {
+    ApiAuditEvent event = getTimeCheckedEvent();
+    assertNull(event.getGraphqlQueryDetails());
+    assertEquals(null, event.getGraphqlErrorPaths());
```
Comment (jeskew-gov) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java:76]:
```suggestion
    assertNull(event.getGraphqlQueryDetails());
```

File: backend/src/test/java/gov/cdc/usds/simplereport/api/AuditLoggingFailuresTest.java (hunk 5 above)
Comment (benwarfield-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/AuditLoggingFailuresTest.java:89]: No, it should have thrown some￼ custom exception indicating the bad response from the server. I need to file a bug about that. The underlying ￼issue here￼ is that raising an exception in the instrumentation causes a 400 error, which is probably not what we want, but I'm not sure what to do about it ￼

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/AuditLoggingAdvice.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        responseCode = responseStatus.code().value();
+      }
+      PatientLink patientLink = _contextHolder.getPatientLink();
+      Organization org = _contextHolder.getOrganization();
+      LOG.trace(
+          "Saving audit entry for patientLinkId={}, organizationId={}",
+          patientLink.getInternalId(),
+          org.getInternalId());
+      String requestId = MDC.get(LoggingConstants.REQUEST_ID_MDC_KEY);
+      _auditService.logRestEvent(requestId, request, responseCode, org, patientLink);
```
Comment (nullflux) [inline backend/src/main/java/gov/cdc/usds/simplereport/logging/AuditLoggingAdvice.java:58]: ✅

File: backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java (hunk 6 above)
Comment (jeskew-gov) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java:76]: Oops, I read this as `assertNull(X); assertEquals(null, X);` disregard

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Override
+  public void postHandle(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      Object handler,
+      ModelAndView modelAndView)
+      throws Exception {
+    // this turns out not to be a place where we can modify the response status in case of failures,
+    // so we had to move that service call somewhere else
```
Comment (nullflux) [inline backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:48]: Could you update this comment to identify where we're doing that?

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/RestAuditLogManager.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    super();
+    this._auditService = auditService;
+    this._contextHolder = contextHolder;
+  }
+
+  public boolean logRestSuccess(HttpServletRequest request, Object resultObject) {
+    PatientLink patientLink = _contextHolder.getPatientLink();
+    if (patientLink == null) {
+      LOG.error("Somehow reached PXP success handler with no patient link");
+      return false;
```
Comment (nullflux) [inline backend/src/main/java/gov/cdc/usds/simplereport/logging/RestAuditLogManager.java:38]: is there any impact on the annotated controller in the event of a false return here? I couldn't immediately find an answer in the spring-security docs, and it's certainly not in the javadoc

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientLinkService.java [hunk 10]
```diff
@@ -66,11 +73,13 @@ public boolean verifyPatientLink(UUID internalId, LocalDate birthDate)
       TestOrder testOrder = patientLink.getTestOrder();
       Person patient = testOrder.getPatient();
 
-      if (testOrder.getPatient().getBirthDate().equals(birthDate)) {
+      if (patient.getBirthDate().equals(birthDate)) {
+        LOG.trace("Successfully authenticated patient {}", patient.getInternalId());
+        contextHolder.setContext(patientLink, testOrder, patient);
         if (patientLink.isExpired()) {
+          LOG.trace("Link is expired!");
```
Comment (nullflux) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientLinkService.java:80]: Thanks for adding these extra trace logs 🙏

**Review by nullflux — APPROVED** [review]
Late to the party but this is ✅  from me! Couple of questions but nothing blocking below

File: backend/src/main/java/gov/cdc/usds/simplereport/service/errors/RestAuditFailureException.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+/**
+ * An error indicating that the audit logging step failed, and should not be retried. Since we do
+ * not care to disclose this information, and the most likely cause is a maliciously crafted request
+ * (unless the database went down at a very, very precise moment in time), we instruct the servlet
+ * to respond with {@link HttpStatus#BAD_REQUEST}.
+ */
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
Comment (jeskew-gov) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/errors/RestAuditFailureException.java:12]: Super nitpicky and ignorable -- shouldn't this be a 500?

File: backend/src/test/java/gov/cdc/usds/simplereport/test_util/TestDataFactory.java [hunk 12]
```diff
@@ -39,6 +39,11 @@
 @Component
 public class TestDataFactory {
 
+  // actually Fred Astaire's birthday! Not random, just a little arbitrary.
```
Comment (jeskew-gov) [inline backend/src/test/java/gov/cdc/usds/simplereport/test_util/TestDataFactory.java:42]: 😄

**Review by jeskew-gov — APPROVED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/api/pxp/PatientExperienceController.java [hunk 13]
```diff
@@ -36,12 +38,15 @@
...
- * documentation on both sides
+ * Note that this controller self-authorizes by means of this {@link PreAuthorize} annotation and
+ * its routes are set to permitAll() in SecurityConfiguration. If this changes, please update the
+ * documentation on both sides.
+ *
+ * <p>Similarly, this controller sends information to the audit log via a {@link PostAuthorize}
  */
 @PreAuthorize(
     "@patientLinkService.verifyPatientLink(#body.getPatientLinkId(), #body.getDateOfBirth())")
+@PostAuthorize("@restAuditLogManager.logRestSuccess(#request, returnObject)")
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/pxp/PatientExperienceController.java:49]: This requires having an HttpServletRequest argument to every method, which I don't love, but it's a lot less magical than the other options.

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java (hunk 4 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java:57]: LOL turns out I was right about it not working in afterCompletion, but wrong about it working in postHandle. &lt;sigh&gt;

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/RestAuditLogManager.java (hunk 9 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/RestAuditLogManager.java:38]: It will convert the response to a 403, I believe. So yes. 😬

File: backend/src/main/java/gov/cdc/usds/simplereport/service/errors/RestAuditFailureException.java (hunk 11 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/errors/RestAuditFailureException.java:12]: I went with 400 on the grounds that the most likely reason for a failure was some manner of intentionally-crafted nutty request. That is admittedly speculative on my part, though—it definitely _is_ an internal server error, so 500 (or rather, leaving it out and letting the servlet default to that) was my initial idea. I'm not convinced either way right now.

File: backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java (hunk 6 above)
Comment (benwarfield-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java:76]: I mean, the `assertEquals(null, Y)` still ought to be fixed, it's just a slightly different fix. 🙂

File: backend/src/test/java/gov/cdc/usds/simplereport/api/pxp/PatientExperienceControllerTest.java [hunk 14]
```diff
@@ -79,13 +78,17 @@ void preAuthorizerThrows403() throws Exception {
         "{\"patientLinkId\":\"" + UUID.randomUUID() + "\",\"dateOfBirth\":\"" + dob + "\"}";
 
     MockHttpServletRequestBuilder builder =
-        put("/pxp/link/verify")
+        put(VERIFY_LINK)
```
Comment (benwarfield-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/pxp/PatientExperienceControllerTest.java:81]: Will pull these into a central location sometime later this week or next.

File: backend/src/main/java/gov/cdc/usds/simplereport/logging/AuditLoggingAdvice.java (hunk 7 above)
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/logging/AuditLoggingAdvice.java:58]: I just was staring at this wondering why I didn't catch the possible exception, and then I remembered this was the exception handler. 🤦
